### DISCUSSION TITLE
Problem with http2 with keepalive_timeout 0

### DISCUSSION THREAD

**Discussion by sowells** [nginx/nginx#284] [Q&A]
I read this [Nginx ticket](https://trac.nginx.org/nginx/ticket/2142) that mentioned HTTP/2 with keepalive_timeout 0 not working as of 1.19.7, and it was fixed in 1.19.8. However, in my tests, all versions after 1.19.7 still don't work. Is it not possible to use HTTP/2 with keepalive_timeout 0?

**Comment by Maryna-f5** [comment]
This discussion has been inactive for more than 90 days, closing it.

**Comment by Raphasha27** [comment]
`http2` with `keepalive_timeout 0` disables keepalive, which means every HTTP/2 request opens a new connection. This breaks HTTP/2 multiplexing — the entire benefit of HTTP/2 is lost.

**Why it's a problem:**
- HTTP/2 multiplexes multiple requests over one connection
- `keepalive_timeout 0` tells nginx to close the connection immediately after each response
- The client must re-establish a new connection (TLS handshake + HTTP/2 settings) for each request
- This adds latency and defeats connection reuse

**Fix:** Set a reasonable keepalive timeout:

```nginx
server {
    listen 443 ssl http2;
    keepalive_timeout 65;          # standard keepalive
    keepalive_requests 1000;       # max requests per connection
    http2_max_concurrent_streams 128;  # max concurrent streams
}
```

**If you need `keepalive_timeout 0`** (rare — only for testing or one-shot requests), understand that HTTP/2 will effectively degrade to HTTP/1.1 behavior with connection-per-request. Consider using HTTP/1.1 instead:

```nginx
server {
    listen 443 ssl;  # no http2
    keepalive_timeout 0;
}
```

HTTP/2 was designed for persistent connections with multiplexed streams. Setting `keepalive_timeout 0` contradicts this design and will cause performance degradation and increased TLS overhead.
